### PR TITLE
feat(dflash): add sampled-anchor walk metrics

### PR DESCRIPTION
**Pull request by maocheng23**
## Motivation

Training averages over sampled draft blocks, while serving skips intervening anchors after accepting a prefix. Add the average accepted length over the sampled anchors that this walk actually visits.

This is a focused alternative to #835, originally proposed by @lvhua898-cpu with follow-up contributions by @jiapingW. Both contributors are credited as co-authors. Sparse anchor sampling remains a proxy for serving, rather than a full SGLang generation evaluation.

## Modifications

- Add all five walk ratios: `dflash/hard_label`, `dflash2/selector`, `domino/final`, `domino/base`, and `dspark/hard_label`, each under `/walk_accepted_length`.
- Reconstruct accepted prefixes across objective chunks and walk once in token-position order. Use the complete masks already held by `forward` for block validity.
- Reuse unary predictions, the existing DFlash2 serving path, and teacher-forced Domino/DSpark predictions. For causal greedy heads, histories agree up to the first rejected token, which is where the accepted prefix stops. No additional Domino GRU replay or DSpark Markov sampling is needed.
- Collect the new ratios on detailed-metric steps and use the existing distributed numerator/denominator reduction. Preserve losses, gradients, and existing per-block metrics.

Production changes are **+149/-4 lines** in `dflash_family_model.py` and **8 added lines** in the strategies, plus one 281-line parameterized test module. The shared chunk reducer and model heads are unchanged.

## Related Issues

- #835 — proposed alternative implementation with the same five walk metrics.

## Accuracy Test

```bash
OMP_NUM_THREADS=2 python -m pytest -q \
  tests/test_utils/test_dflash_walk.py \
  tests/test_utils/test_dflash_losses.py \
  tests/test_utils/test_chunking.py \
  tests/test_utils/test_domino_cross_entropy.py \
  tests/test_modeling/test_domino_model.py \
  tests/test_modeling/test_dflash2.py \
  tests/test_modeling/test_draft_registry.py \
  tests/test_runtime/test_domino_launch.py \
  tests/test_runtime/test_dspark_disagg_launch.py \
  tests/test_runtime/test_dspark_offline_launch.py \
  tests/test_runtime/test_seam_fixes.py
```

**121 passed, 11 skipped, 8 subtests passed** on CPU with PyTorch 2.13.0. Coverage includes deterministic walk numerators/denominators, chunk sizes 0/1/2/4/5, gradient parity, detailed-metric gating through strategies, masked holes, sparse/padded/empty anchor masks, and float32/BF16 causal-prefix equivalence using real Domino and DSpark heads (including vanilla, gated, and RNN Markov variants).

An additional 120-case comparison across model families, chunk sizes, and full/holey/empty-row masks matched all five walk ratios from #835 at `0bb4c82e` and preserved losses and existing metrics from main at `13e11dcb`.

Not executed: CUDA/Triton kernel validation or multi-GPU training. The skips include hardware-only tests and the causal-head comparison cases that do not apply to DFlash/DFlash2.

## Benchmark & Profiling

No model-scale throughput benchmark was run. New telemetry adds strict-prefix calculations, small per-anchor tensors, and a CPU walk on detailed-metric steps; it adds no vocabulary projection or autoregressive head replay.

## Checklist

- [x] `pre-commit run --all-files` passed.
- [x] Added focused parameterized tests and ran the relevant existing suites.
- [x] Documented sampled-anchor semantics and validation limits.

### CODE REVIEW

**Review by jiapingW — APPROVED** [review]
LGTM.
